### PR TITLE
#825 - Add find_target_device_by_pid utility method to the SupernovaI3CBlockingInterface class

### PR DESCRIPTION
**Pull request by Salonso1602**
# Resolves  
[825](https://focusuy.atlassian.net/browse/BMC2-825)  

# How to test  
- Plug in Supernova (with 2.0.0 fw)  
- Plug in ICM42605 I3C target
- run examples: (recommended to plug and unplug supernova between executions)  
    - Basic I3C
    - IBI
    - ICM42605

# What to expect  
### basic i3c  
```
Opening Supernova host adapter device and getting access to the I3C protocol interface...
Initializing the bus...

Targets in the I3C bus: {'static_address': 0, 'dynamic_address': 8, 'bcr': 39, 'dcr': 160, 'pid': ['0x00', '0x00', '0x00', '0x00', '0x6a', '0x04']}
Target address: 8 

CCC Transfers
GETPID: [4, 106, 0, 0, 0, 0] 

GETBCR: 27 

GETDCR: A0 

GETCAPS (Get Capabilities): None 

GETMXDS (Get Max Data Speed): None 

GETMRL (Get Max Read Length): 16

UNICAST SETMRL (Set Max Read Length) in: 256 

GETMRL: 256

GETMWL (Get Max Write Length): 8 

BROADCAST SETMWL (Set Max Write Length) in: 128

GETMWL: 128

Write/Read transfers
Write [64] in subaddress 22
Read from subaddress 22: [64]

Write [222, 173, 190, 239] in subaddress 22
Read from subaddress 22: [222, 173, 190, 239]

Reset bus
Resetting bus...
Targets in the I3C bus: []  
```

### IBI  
```
{'static_address': 0, 'dynamic_address': 8, 'bcr': 39, 'dcr': 160, 'pid': ['0x00', '0x00', '0x00', '0x00', '0x6a', '0x04']}
[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

###   ICM42605  
```
{'static_address': 0, 'dynamic_address': 8, 'bcr': 39, 'dcr': 160, 'pid': ['0x00', '0x00', '0x00', '0x00', '0x6a', '0x04']}
Who Am I Register:  [66]
Power Management Register:  [0]
Power Management Register written.
Gyro Config Register:  [6]
Gyro Config Register written.
Status Register:  [0]  
... *Returns a bunch of values  

```

### CODE REVIEW

**Review by gusaaaaa — DISMISSED** [review]
Good work. Let's add some specs to tests.

File: examples/basic_i3c_example.py [hunk 1]
```diff
@@ -34,17 +27,15 @@ def main():
         print("I couldn't initialize the bus. Are you sure there's any target connected?")
         exit(1)
 
-    (_, targets) = i3c.targets()
-
     # Target PID in hexadecimal format
     target_pid = [0x00, 0x00, 0x00, 0x00, 0x6A, 0x04]
```
Comment (pons93) [inline examples/basic_i3c_example.py:31]: This `target_pid` is reversed. The correct pid for the ICM42605 is [04-6A-00-00-00-00].

File: examples/ICM42605_i3c_example.py [hunk 2]
```diff
@@ -26,17 +19,15 @@ def main():
         print("I couldn't initialize the bus. Are you sure there's any target connected?")
         exit(1)
 
-    (_, targets) = i3c.targets()
-
     # Target PID in hexadecimal format
     target_pid = [0x00, 0x00, 0x00, 0x00, 0x6A, 0x04]
```
Comment (pons93) [inline examples/ICM42605_i3c_example.py:23]: This `target_pid` is reversed. The correct pid for the ICM42605 is [04-6A-00-00-00-00].

File: examples/ibi_example.py [hunk 3]
```diff
@@ -25,17 +18,15 @@ def main():
         print("I couldn't initialize the bus. Are you sure there's any target connected?")
         exit(1)
 
-    (_, targets) = i3c.targets()
-
     # Target PID in hexadecimal format
     target_pid = [0x00, 0x00, 0x00, 0x00, 0x6A, 0x04]
```
Comment (pons93) [inline examples/ibi_example.py:22]: This `target_pid` is reversed. The correct pid for the ICM42605 is [04-6A-00-00-00-00].

File: supernovacontroller/sequential/i3c.py [hunk 4]
```diff
@@ -255,6 +255,48 @@ def targets(self):
...
+                - 'pid': Unique ID (Provisional ID) containing a manufacturer ID, a part ID and an instance ID.
+        """
+        try:
+            responses = self.controller.sync_submit([
+                lambda id: self.driver.i3cGetTargetDeviceTable(id)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        filtered_devices = list(filter(lambda device: device["pid"][::-1] == pid, responses[0]["table"]))
```
Comment (pons93) [inline supernovacontroller/sequential/i3c.py:281]: The reversion of this PID is not longer necessary for new SDK 2.X.X series.

File: supernovacontroller/sequential/i3c.py [hunk 5]
```diff
@@ -255,6 +255,48 @@ def targets(self):
...
+
+        if (len(filtered_devices) == 0):
+            return (False, None)
+        
+        target_info = filtered_devices[0]
+        static_address = target_info["staticAddress"]
+        dynamic_address = target_info["dynamicAddress"]
+        bcr = int(target_info["bcr"]["value"][2][2:4], 16)
+        dcr = target_info["dcr"]
+        pid = target_info["pid"][::-1] # Reversing using list slicing
```
Comment (pons93) [inline supernovacontroller/sequential/i3c.py:291]: The reversion of this PID is not longer necessary for new SDK 2.X.X series.

**Review by pons93 — CHANGES_REQUESTED** [review]
Nice and elegant job!!!

I just left a comment regarding the no longer necessity for reversing the PID.

**Review by pons93 — APPROVED** [review]
LGTM!!!!!

Just remember that `SupernovaController` v1.1.0 was already released and v1.2.0 is about to be released. Maybe consider to changing base to develop-1.3.0.

**Comment by Salonso1602** [comment]
> LGTM!!!!!
> 
> Just remember that `SupernovaController` v1.1.0 was already released and v1.2.0 is about to be released. Maybe consider to changing base to develop-1.3.0.

Done!

**Pull request merged**